qt/Torrent.cc

```diff
@@ -173,228 +108,262 @@ Torrent::KeyList const Torrent::detailStatKeys{
+        setme = value;
     }
 
     return changed;
 }
 
-bool Torrent::setDouble(int i, double value)
+bool change(double& setme, tr_variant const* value)
 {
-    bool changed = false;
-
-    assert(0 <= i && i < PROPERTY_COUNT);
-    assert(myProperties[i].type == QVariant::Double);
-
-    if (myValues[i] != value)
-    {
-        myValues[i].setValue(value);
-        changed = true;
-    }
-
-    return changed;
+    double d;
+    return tr_variantGetReal(value, &d) && change(setme, d);
 }
 
-bool Torrent::setTime(int i, time_t value)
+bool change(int& setme, tr_variant const* value)
 {
-    bool changed = false;
+    int64_t i;
+    return tr_variantGetInt(value, &i) && change(setme, static_cast<int>(i));
+}
 
-    assert(0 <= i && i < PROPERTY_COUNT);
-    assert(myProperties[i].type == QVariant::DateTime);
+bool change(uint64_t& setme, tr_variant const* value)
+{
+    int64_t i;
+    return tr_variantGetInt(value, &i) && change(setme, static_cast<uint64_t>(i));
+}
 
-    auto& oldval = myValues[i];
-    auto const newval = qlonglong(value);
+bool change(time_t& setme, tr_variant const* value)
+{
+    int64_t i;
+    return tr_variantGetInt(value, &i) && change(setme, static_cast<time_t>(i));
+}
 
-    if (oldval != newval)
-    {
-        oldval = newval;
-        changed = true;
-    }
+bool change(Speed& setme, tr_variant const* value)
+{
+    int64_t i;
+    return tr_variantGetInt(value, &i) && change(setme, Speed::fromBps(i));
+}
 
-    return changed;
+bool change(bool& setme, tr_variant const* value)
+{
+    bool b;
+    return tr_variantGetBool(value, &b) && change(setme, b);
 }
 
-bool Torrent::setSize(int i, qulonglong value)
+bool change(QString& setme, tr_variant const* value)
 {
     bool changed = false;
+    char const* str;
+    size_t len;
 
-    assert(0 <= i && i < PROPERTY_COUNT);
-    assert(myProperties[i].type == QVariant::ULongLong);
+    if (!tr_variantGetStr(value, &str, &len))
+    {
+        return changed;
+    }
 
-    if (myValues[i].isNull() || myValues[i].toULongLong() != value)
+    if (len == 0)
     {
-        myValues[i].setValue(value);
-        changed = true;
+        changed = !setme.isEmpty();
+        setme.clear();
+        return changed;
     }
 
-    return changed;
+    return change(setme, QString::fromUtf8(str, len));
 }
 
-bool Torrent::setString(int i, char const* value, size_t len)
+bool change(Peer& setme, tr_variant* value)
 {
     bool changed = false;
 
-    assert(0 <= i && i < PROPERTY_COUNT);
-    assert(myProperties[i].type == QVariant::String);
-
-    auto& oldval = myValues[i];
-    auto const newval = QString::fromUtf8(value, len);
-
-    if (oldval != newval)
+    size_t pos = 0;
+    tr_quark key;
+    tr_variant* child;
+    while (tr_variantDictChild(value, pos++, &key, &child))
     {
-        oldval = newval;
-        changed = true;
+        switch (key)
+        {
+#define HANDLE_KEY(key) case TR_KEY_ ## key: \
+    changed |= change(setme.key, child); break;
```

Comment:
Here and elsewhere: not a big fan of bitwise operations on non-integer values. I'm always for proper boolean expressions (in C as well as in C++), so consider
```suggestion
    changed = change(setme.key, child) || changed; break;
```
or
```suggestion
    if (change(setme.key, child)) \
    { \
        changed = true; \
    } \
    break;
```

Reply:
I'd push back on this a little bit: the first suggestion to me is more error-prone since one has to repeat the variable name and is more complicated to read. The second alternative has simpler logic but is prolix, replacing a two-character `|=` operator with a new paragraph.

If you really feel strongly about this I'll replace the code with the second option, but wanted to ask first -- IMO `|=` is a clear, concise way of expressing this logical operation

Reply:
The point is that `|=` is a bitwise, not a logical operation, and it's basically the only reason I'm against it; if there was a logical counterpart for it (say, `||=`), replacing one with the other wouldn't be equivalent (due to short-circuiting guarantee) which may be confusing. I don't see an issue with variable name being repeated as this is exactly why you're defining a macro (so that it's repeated once and not for every field).

That's how I feel about it, but I certainly see the appeal of writing it the way you did. The fact is just that if I were to write this code I would do it differently. Whether that's a good reason to change it is an open question, and that's why I'm deliberately saying  "consider" in such cases where I'm not ever going to insist.

Reply:
That's reasonable. Changed in 7739261fced79ad761f515a568f03343519c2b29